msm/msm

```diff
@@ -242,41 +344,145 @@ function update() {
   done
 }
 
+OPT=$1
+shift
 
-function search() {
-  search_list=$(list | grep 'submodule "' | sed 's/\[submodule "//g'| sed 's/"\]//g')
-  while [[ $# -gt 0 ]] ; do
-    search_string=$1
-    shift
-    while read -r matches; do
-      if [[ "${search_string}" == "${matches}" ]] ; then
-        echo "Exact match found: ${matches}"
+case ${OPT} in
+  "install")
+      if [[ $# -gt 0 ]] ; then
+         get_skill_list
+         exit_code=$?
+         if [[ ${exit_code} -gt 0 ]] ; then
+             echo "${script}: error ${exit_code}"
+             exit ${exit_code}
+         fi
+        
+         for str in "$@"
+         do
+            install $str
+            rc=$?
+
+            if [[ ${rc} -gt 0 ]] ; then
+               if [[ ${rc} -gt ${exit_code} ]] ; then
+                  exit_code=${rc}
+               fi
+            fi
+         done
       else
-        echo "Possible match: ${matches}"
+         # install requires a parameter, show help
+         help
+         exit_code=1
+      fi
+      ;;
+  "remove")
+      if [[ $# -gt 0 ]] ; then
+         get_skill_list
+         exit_code=$?
+         if [[ ${exit_code} -gt 0 ]] ; then
+             echo "${script}: error ${exit_code}"
+             exit ${exit_code}
+         fi
+         
+         for str in "$@"
+         do
+            remove $str
+            rc=$?
+
+            if [[ ${rc} -gt 0 ]] ; then
+               if [[ ${rc} -gt ${exit_code} ]] ; then
+                  exit_code=${rc}
+               fi
+            fi
+         done
+      else
+         # remove requires a parameter, show help
+         help
+         exit_code=1
+      fi
+      ;;
+  "list")
+      get_skill_list
+      exit_code=$?
+      if [[ ${exit_code} -gt 0 ]] ; then
+          echo "${script}: error ${exit_code}"
+          exit ${exit_code}
+      fi
+      echo "${LIST_CACHE}" | grep 'submodule "' | sed 's/\[submodule "//g'| sed 's/"\]//g' | sort
+      exit_code=$?
+      ;;
+  "update")
+      get_skill_list
+      exit_code=$?
+      if [[ ${exit_code} -gt 0 ]] ; then
+         echo "${script}: error ${exit_code}"
+         exit ${exit_code}
       fi
-    done < <(grep -i "${search_string}" <<< "${search_list}")
-  done
-}
-
 
-######################################################################
-## Main program
-######################################################################
+      update
+      exit_code=$?
+      ;;
+  "default")
+      echo "=== Checking for default skills"
+      INSTALLING_DEFAULTS="true"
+      get_skill_list
+      exit_code=$?
+      if [[ ${exit_code} -gt 0 ]] ; then
+         echo "${script}: error ${exit_code}"
+         exit ${exit_code}
```

Comment:
These error printout/exit could be moved to a separate function since they're duplicated for each msm command

Reply:
I'm not clear on what you mean here

Reply:
Nothing really important but one could create a function wrapping the above lines since they exist a couple of times in the script.

```bash
function exit_error() {
       echo "$1: error $2"
       exit $2
}